src/iop/negadoctor.c

```diff
@@ -0,0 +1,1387 @@
+  return fmaxf(fmaxf(vector[0], vector[1]), vector[2]);
+}
+
+static inline float v_minf(const float vector[3])
+{
+  // Find the min over an RGB vector
+  return fminf(fminf(vector[0], vector[1]), vector[2]);
+}
+
+
+DT_MODULE_INTROSPECTION(1, dt_iop_negadoctor_params_t)
+
+
+typedef enum dt_iop_negadoctor_filmstock_t
+{
+  // What kind of emulsion are we working on ?
+  DT_FILMSTOCK_NB = 0,
+  DT_FILMSTOCK_COLOR = 1,
+  DT_FILMSTOCK_END = 2
+} dt_iop_negadoctor_filmstock_t;
+
+
+typedef enum dt_iop_negadoctor_pickcolor_type_t
+{
+  DT_NEGADOCTOR_NONE = 0,
+  DT_NEGADOCTOR_DMIN = 1,
+  DT_NEGADOCTOR_WB_high = 2,
+  DT_NEGADOCTOR_OFFSET = 3,
+  DT_NEGADOCTOR_DMAX = 4,
+  DT_NEGADOCTOR_WB_low = 5,
+  DT_NEGADOCTOR_EXPOSURE = 6,
+  DT_NEGADOCTOR_BLACK = 7
+} dt_iop_negadoctor_pickcolor_type_t;
+
+
+typedef struct dt_iop_negadoctor_params_t
+{
+  dt_iop_negadoctor_filmstock_t film_stock;
+  float DT_ALIGNED_PIXEL Dmin[4];         // color of film substrate
+  float DT_ALIGNED_PIXEL wb_high[4];      // white balance RGB coeffs (illuminant)
+  float DT_ALIGNED_PIXEL wb_low[4];       // white balance RGB offsets (base light)
+  float D_max;                            // max density of film
+  float offset;                           // inversion offset
+  float black;                            // display black level
+  float gamma;                            // display gamma
+  float soft_clip;                        // highlights roll-off
+  float exposure;                         // extra exposure
+} dt_iop_negadoctor_params_t;
+
+
+typedef struct dt_iop_negadoctor_data_t
+{
+  float DT_ALIGNED_PIXEL Dmin[4];         // color of film substrate
+  float DT_ALIGNED_PIXEL wb_high[4];      // white balance RGB coeffs / Dmax
+  float DT_ALIGNED_PIXEL offset[4];       // inversion offset
+  float black;                            // display black level
+  float gamma;                            // display gamma
+  float soft_clip;                        // highlights roll-off
+  float soft_clip_comp;                   // 1 - softclip, complement to 1
+  float exposure;                         // extra exposure
+} dt_iop_negadoctor_data_t;
+
+
+typedef struct dt_iop_negadoctor_gui_data_t
+{
+  GtkNotebook *notebook;
+  GtkWidget *film_stock;
+  GtkWidget *Dmin_R, *Dmin_G, *Dmin_B;
+  GtkWidget *wb_high_R, *wb_high_G, *wb_high_B;
+  GtkWidget *wb_low_R, *wb_low_G, *wb_low_B;
+  GtkWidget *D_max;
+  GtkWidget *offset;
+  GtkWidget *black, *gamma, *soft_clip, *exposure;
+  GtkWidget *Dmin_picker, *Dmin_sampler;
+  GtkWidget *WB_high_picker, *WB_high_sampler;
+  GtkWidget *WB_low_picker, *WB_low_sampler;
+  dt_iop_color_picker_t color_picker;
+} dt_iop_negadoctor_gui_data_t;
+
+
+typedef struct dt_iop_negadoctor_global_data_t
+{
+  int kernel_negadoctor;
+} dt_iop_negadoctor_global_data_t;
+
+
+const char *name()
+{
+  return _("negadoctor");
+}
+
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_ALLOW_TILING;
```

Comment:
Does it makes sense to have multiple instances ?

Reply:
No, indeed, I didn't thought of it.